pegin/getaddress.go

```diff
@@ -0,0 +1,379 @@
+)
+
+type NetworkType int
+
+type AddressInfo struct {
+	ClaimScript      string
+	MainChainAddress string
+}
+
+// GetAddressInfo simulates getpeginaddress RPC call which returns mainChain
+// address to which BTC should be sent in BTC network and claim script that is
+// to be used in claimpegin RPC alongside with bitcoinTx and txoutproof
+func GetAddressInfo(
+	privateKey btcec.PrivateKey,
+	fedpegInfo FedpegInfo,
+	networkType NetworkType,
+	isDynaFedEnabled bool,
+) (*AddressInfo, error) {
+	liquidNetwork, btcNetwork := getNetworkParams(networkType)
+
+	p2wkh, witnessScript := getClaimScriptDetails(privateKey, liquidNetwork)
+
+	contract, err := calculateContract(
+		fedpegInfo.FedpegScript,
+		witnessScript,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	mainChainAddress, err := createMainChainAddress(
+		contract,
+		btcNetwork,
+		isDynaFedEnabled,
+		fedpegInfo.FedpegScript,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &AddressInfo{
+		ClaimScript:      p2wkh,
+		MainChainAddress: mainChainAddress,
+	}, nil
+}
+
+func getNetworkParams(
+	networkType NetworkType,
+) (*network.Network, *chaincfg.Params) {
+	var liquidNetwork *network.Network
+	var btcNetwork *chaincfg.Params
+	switch networkType {
+	case MainNet:
+		liquidNetwork = &network.Liquid
+		btcNetwork = &chaincfg.MainNetParams
+	case RegtestNet:
+		liquidNetwork = &network.Regtest
+		btcNetwork = &chaincfg.RegressionNetParams
+	}
+
+	return liquidNetwork, btcNetwork
+}
+
+func getClaimScriptDetails(
+	privateKey btcec.PrivateKey,
+	net *network.Network,
+) (string, []byte) {
+	p2wpkh := payment.FromPublicKey(
+		privateKey.PubKey(),
+		net,
+		nil,
+	)
+	return hex.EncodeToString(p2wpkh.WitnessScript), p2wpkh.WitnessScript
+}
+
+func createMainChainAddress(
+	contract []byte,
+	btcNetwork *chaincfg.Params,
+	isDynaFedEnabled bool,
+	fedpegScript []byte,
+) (string, error) {
+	pops, err := address.ParseScript(fedpegScript)
+	if err != nil {
+		return "", err
+	}
+
+	var mainChainAddress string
+	if isDynaFedEnabled || address.IsScriptHash(pops) {
+		//P2SH - P2WSH(P2CH)
+		witnessScriptHash := sha256.Sum256(contract)
+		builder := txscript.NewScriptBuilder()
+		builder.AddOp(txscript.OP_0).AddData(witnessScriptHash[:])
+		script, err := builder.Script()
+		if err != nil {
+			return "", err
+		}
+		ps2h, err := btcutil.NewAddressScriptHash(script, btcNetwork)
+		if err != nil {
+			return "", err
+		}
+		mainChainAddress = ps2h.String()
+	} else {
+		//P2WSH - P2CH
+		witnessScriptHash := sha256.Sum256(contract)
+		p2wsh, err := btcutil.NewAddressWitnessScriptHash(
+			witnessScriptHash[:],
+			btcNetwork,
+		)
+		if err != nil {
+			return "", err
+		}
+		mainChainAddress = p2wsh.String()
+	}
+
+	return mainChainAddress, nil
+}
+
+type FedpegInfo struct {
+	FedpegScript  []byte
+	FedpegProgram []byte
+}
+
+// TODO - get fedpeg details from live blockchain
+func GetFedpegInfo() (*FedpegInfo, error) {
+	return &FedpegInfo{}, nil
+}
+
+func calculateContract(
```

Comment:
this method seems the one maybe to be split, so other targets could use their own secp. Maybe we should decompose calculate contract in many little functions that only do data structure manipulation and expect data as input?

Like, generateContract(watchman, tweaked, ...) 


Any other idea?

Reply:
As agreed, contract func extracted to other pkg